bazaarvoice/emodb
# Modified Stash to only scan over "live" token ranges

## Github Issue #

None

## What Are We Doing Here?

Currently stash works by querying the entire token range and converting data to JSON files as it is read. Because of this all data is queried, including rows from deleted tables. Any rows found from deleted tables are skipped by the stash process. This works under the following assumptions:

1. The Cassandra delta table is dominated by data from "live" (undeleted) tables.
2. Table deletions are rare.
3. There are no issues scanning over records for a deleted table which have been purged.

While these may have been true when Stash was first developed, over time these have become incorrect assumptions:

1. In our experience deletes, particularly of large tables, are common as teams develop and test new features.
2. Because of tombstones scanning over large swaths of purged tables can exceed tombstone limit thresholds set by Cassandra.

This PR changes stash to only operate on "live" token ranges -- ranges which are known to belong to shards for undeleted tables.  These changes include the following:

1. Prior to Stash starting a snapshot is taken for all "live" token ranges.
2. When a broad Stash token range is scanned it splits this into a scan for each "live" token range which falls within the encompassing range.

## How to Test and Verify

1. Check out this PR.
2. Run Stash against an exiting Emo deployment and verify the Stash data is consistent with the output from an existing Stash.
3. Additionally to test locally use `ScanUploadTest`.

## Risk

The risk to normal Emo services is low since the affected code is limited to Stash.  However, the risk for Stash is high since an error in this PR could lead to token ranges of accessibly data being skipped by Stash.

### Level 

`High` The changes aren't too large or complicated, but because Stash is a core component of Emo the level is naturally higher.

### Required Testing

`Regression`, with some `Manual` warranted to validate Stash contents

### Risk Summary

One unknown which is difficult to test in a sandbox is whether the addition of multiple subqueries will significantly slow Stash down.  It doesn't appear that this should be the case since the additional round-trips for queries are likely on par with existing query fetch sizes, and even if not there are additional parallel threads resolving and uploading Stash files which are likely a larger bottleneck than the raw Cassandra queries, but whether this analysis is correct or not needs to be proven.

## Code Review Checklist


build and run.

we inevitably come back to this code it will only take hours to figure out, not
days.

a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.
